Enable Select pushdown on uncompressed files

preethiratnam opened the pull request:
## Description

This PR fixes issue #2475 . We have identified the root cause to be a problem with the field delimiter passed in the S3 Select query. The Hive connector supplies a default field delimiter of [byte 1](https://www.fileformat.info/info/unicode/char/0001/index.htm) when it is not explicitly set in the Hive table. This is configured in [LazySerDeParameters](https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L51). The byte 1 value was getting passed as a String type "1" in the [S3SelectCsvRecordReader](https://github.com/trinodb/trino/blob/fe608f2723842037ff620d612a706900e79c52c8/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java#L67-L73).

In this case, the query to S3 Select used "1" as the field delimiter and caused unexpected results. This bug is present in both uncompressed and compressed file formats. It was detected in uncompressed files because the test input (test_table.csv) contained "1".  However, the same issue can be reproduced with compressed files as well.

To fix this, I have removed the default field delimiter from S3SelectCsvRecordReader. We will now use the field delimiter from the schema only if it is explicitly set. In other cases, we will not pass any field delimiter, which defaults to a comma in S3 Select. This is a reasonable default for CSV file formats.

> Is this change a fix, improvement, new feature, refactoring, or other?

Fix.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Hive S3 Select connector

> How would you describe this change to a non-technical end user or system administrator?

Enables S3 Select for uncompressed files. Fixes a bug in S3 Select queries.

## Related issues, pull requests, and links

* Fixes #2475

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix and reenable S3 Select for uncompressed files.
```

findepi commented:
@preethiratnam can you please check CI results?

preethiratnam commented:
> @preethiratnam can you please check CI results?

Thank you, I've fixed the failing test now. Can you please re-run them?

preethiratnam commented:
> Thank you for your pull request and welcome to our community. We require contributors to sign our [Contributor License Agreement](https://github.com/trinodb/cla/raw/master/Trino%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. In order for us to review and merge your code, please submit the signed CLA to [cla@trino.io](mailto:cla@trino.io). For more information, see https://github.com/trinodb/cla.

I emailed my signed CLA on 1-Jun, how long does it take to get updated? Do I need to raise a PR on https://github.com/trinodb/cla/blob/master/contributors? @findepi

preethiratnam commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java:
```diff
@@ -224,7 +224,7 @@ public float getProgress()
-    String getFieldDelimiter(Properties schema)
+    protected String getFieldDelimiter(Properties schema)
     {
         return schema.getProperty(FIELD_DELIM, schema.getProperty(SERIALIZATION_FORMAT));
```
Good point. I think as Select pushdown only support CSV files at the moment, this method will always be overridden. It's possibly useful for other file formats in the future, so I left it as is.

findepi replied:
What's an example value of the `schema.getProperty(SERIALIZATION_FORMAT)` ?

preethiratnam replied:
`schema.getProperty(SERIALIZATION_FORMAT)` has default separators of single byte characters eg: byte 1, byte 2 ([code](https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L51)).

findepi commented:
@preethiratnam i posted https://github.com/trinodb/trino/pull/12702 to run the tests with secrets. please let me know if this is green too

findepi commented:
Test PR with secrets: https://github.com/trinodb/trino/pull/12702
(cc @ilfrin @nineinchnick for potentially streamlining this)

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java:
```diff
@@ -91,4 +92,11 @@ public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, St
     }
+
+    @Override
+    protected String getFieldDelimiter(Properties schema)
```
It's kind of weird to override this method, if the only use place is in the `buildSelectObjectRequest` which is also overridden. 
Remove `io.trino.plugin.hive.s3select.S3SelectLineRecordReader#getFieldDelimiter`  implementation, as it's no longer used. Remove the definition, since it's redundant

preethiratnam replied:
Makes sense, I've removed this method in the latest commit.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java:
```diff
@@ -91,4 +92,11 @@ public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, St
+    @Override
+    protected String getFieldDelimiter(Properties schema)
+    {
+        // Use the field delimiter only if it is specified in the schema. If not, use null (defaults to ',' in S3 Select).
```
What does it mean that null defaults to a comma?

preethiratnam replied:
I updated the description in the latest commit. Basically, when you send the field delimiter as `null` on the `SelectObjectContent` request to S3 Select, the S3 Select service uses comma as the field delimiter. Since this class only deals with CSV file formats, it looks to be a reasonable default to me.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java:
```diff
@@ -91,4 +92,11 @@ public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, St
+    @Override
+    protected String getFieldDelimiter(Properties schema)
+    {
+        // Use the field delimiter only if it is specified in the schema. If not, use null (defaults to ',' in S3 Select).
+        return schema.getProperty(FIELD_DELIM);
```
It sounds like we should have S3 Select tests for csv files with various delimiters, not only the default one.

findepi commented:
The change looks input delimiter -related, not compression related, but it supposedly fixes s3 select on uncompressed data.
What do the two (field delim and compression) have in common?

also, please make sure there is test coverage for {compressed, uncompressed} x { default delimter, explicit default delimiter, non-default delimiter }.

preethiratnam commented:
> 

Hi @martint I emailed the CLA on 1-June, but do not see myself listed on the Contributors yet. Can you please check? Thank you.

preethiratnam commented:
> The change looks input delimiter -related, not compression related, but it supposedly fixes s3 select on uncompressed data. What do the two (field delim and compression) have in common?
> 
> also, please make sure there is test coverage for {compressed, uncompressed} x { default delimter, explicit default delimiter, non-default delimiter }.

The issue here was that we saw incorrect results as the field delimiter was passed incorrectly to S3 Select. In the uncompressed file test input `test_table.csv`, we had an input containing "1". This inadvertently helped catch the issue for uncompressed files. The same issue exists in compressed files as well, but we hadn't detected it so far, as none of the compressed file test inputs contained "1".

Since I fixed the issue with input delimiter, we are now seeing correct results for uncompressed files. So I have enabled S3 Select pushdown for uncompressed files as well.

The latest commit has test coverage for all cases. The default delimiter case is already covered by the `testGetRecords` method that was failing previously. I have added tests for explicit default delimiter and non-default delimiter.

@findepi could you please re-run the integration tests? Thank you!

findepi commented:
@preethiratnam sorry, i cannot help with the CLA process.

preethiratnam commented:
Hi @findepi could you please run this PR with tests? I added a new commit yesterday. Thank you!

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java:
```diff
@@ -91,4 +92,11 @@ public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, St
+    protected String getFieldDelimiter(Properties schema)
+    {
+        // Use the field delimiter only if it is specified in the schema. If not, send it as null in the request to S3 Select.
+        // In this case, S3 Select defaults to using ',' as the field delimiter.
+        return schema.getProperty(FIELD_DELIM);
```
if we were to read the file directly, we would need to assume the default delimiter (i guess it's comma, right?)
why not add the defaulting logic here too, so that it remains on the Trino side, and is not up to an external service?

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectPushdown.java:
```diff
@@ -106,9 +106,12 @@ private static boolean isInputFormatSupported(Properties schema)
     {
         if (inputFormat instanceof TextInputFormat) {
-            return getCompressionCodec((TextInputFormat) inputFormat, path)
+            var compressionCodec = getCompressionCodec((TextInputFormat) inputFormat, path);
```
please see https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-var

preethiratnam replied:
Fixed in the latest commit.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectPushdown.java:
```diff
@@ -106,9 +106,12 @@ private static boolean isInputFormatSupported(Properties schema)
+            return compressionCodec.isEmpty() || compressionCodec
                     .map(codec -> (codec instanceof GzipCodec) || (codec instanceof BZip2Codec))
-                    .orElse(false); // TODO (https://github.com/trinodb/trino/issues/2475) fix S3 Select when file not compressed
+                    .orElse(false);
```
`orElse` is useless since you already checked it's not empty

what about removing `compressionCodec.isEmpty() ||` and using `orElse(true)` here?

preethiratnam replied:
Makes sense, fixed in the latest commit.

preethiratnam commented:
@findepi I've addressed your review comments, can you please re-run the tests?

findepi commented:
> [Enable Select pushdown on uncompressed files](https://github.com/trinodb/trino/pull/12633/commits/e2cd8762a509301a01652fc12161f1239f95af16)
>  
> [Fix test as S3 Select now supports pushdown on uncompressed files](https://github.com/trinodb/trino/pull/12633/commits/ab3fe8bad259b73c0b23faec45cf116a56657170)

this should be one commit, right?

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectCsvRecordReader.java:
```diff
@@ -95,8 +96,7 @@ public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, St
     protected String getFieldDelimiter(Properties schema)
     {
-        // Use the field delimiter only if it is specified in the schema. If not, send it as null in the request to S3 Select.
-        // In this case, S3 Select defaults to using ',' as the field delimiter.
```
This should be done right already in the `Enable Select pushdown on uncompressed files` commit

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectPushdown.java:
```diff
@@ -106,12 +106,10 @@ private static boolean isInputFormatSupported(Properties schema)
-            var compressionCodec = getCompressionCodec((TextInputFormat) inputFormat, path);
-
             // S3 Select supports the following formats: uncompressed, GZIP and BZIP2.
-            return compressionCodec.isEmpty() || compressionCodec
```
This should be done right already in the `Enable Select pushdown on uncompressed files` commit

arhimondr commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectPushdown.java:
```diff
@@ -106,9 +106,10 @@ private static boolean isInputFormatSupported(Properties schema)
             return getCompressionCodec((TextInputFormat) inputFormat, path)
                     .map(codec -> (codec instanceof GzipCodec) || (codec instanceof BZip2Codec))
-                    .orElse(false); // TODO (https://github.com/trinodb/trino/issues/2475) fix S3 Select when file not compressed
+                    .orElse(true);
```
question: What if a file is compressed, but with a codec that is not supported? Maybe something like 

```
Optional<Codec> codec = getCompressionCodec((TextInputFormat) inputFormat, path);
if(codec.isEmpty()){
  // assume uncompressed
  return true;
}
```

Also I wonder how safe is to assume that a file is uncompressed if a codec is not found for a given extension? (I guess it is as expected as I see [similar assumptions](https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java#L121) being made in other parts of the code, but want to clarify).

preethiratnam replied:
Hi Andrii, if a file is compressed with a codec that isn't supported, the codec would be different and this would return false:
```
codec -> (codec instanceof GzipCodec) || (codec instanceof BZip2Codec)
```
The default `orElse(true)` is only effective when the codec is null. So the method returns true only when codec is null (uncompressed), Gzip or Bzip2. We also have unit tests for the `isCompressionCodecSupported` method with different codec inputs.

Good point about the null codec assumption, though. I think it's reasonable- if there is no codec defined when a Hive table is created, it doesn't depend on codecs and is expected to have uncompressed files. This method internally uses Hadoop's [CompressionCodecFactory](https://hadoop.apache.org/docs/r2.4.1/api/org/apache/hadoop/io/compress/CompressionCodecFactory.html), which I think is the standard.

Thank you so much for looking into this PR!

arhimondr reviewed (commented):
LGTM % comment
